Handle file naming when HTTP Redirects occur

adamretter opened the pull request:
Before this PR when getting a resource from a URL, if a `<fileOutputName>` was not configured, the resultant file would be named based on the name of the remote resource configured in `<url>`. 

However if the URL leads to a HTTP redirect, then the name of the remote resource changes as we are redirected one or more times. This PR accounts for HTTP redirects, by changing the name of the output file only when:
1. A `<fileOutputName>` is not explicitly configured,
2. and, a HTTP redirect is detected which leads to a differently named resource.

mickaelistria commented on src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java:
```diff
@@ -49,18 +54,42 @@ public HttpFileRequester(final CloseableHttpClient httpClient, final ProgressRep
      * @param clientContext the HTTP execution context.
      */
-    public void download(final URI uri, final File outputFile, final HttpContext clientContext) throws Exception {
+    public File download(final URI uri, final boolean fixedOutputFileName, final File outputFile, final HttpClientContext clientContext) throws Exception {
```
Instead of a boolean, we could try to have a non-null `outputFolder` and a nullable `outputFileName`. if `outputFileName == null` then we run the same code as if `fixedOutputFileName == false`

adamretter replied:
I don't think that will work. As when there is a non-http-redirect, we still want to use the `outputFile`, but with your suggestion that would be `null`.

mickaelistria replied:
If `outputFileName` is set on the Mojo (and passed here) we do want to use it independently of redirects.
In case it's not set, then we can pass the folder and null as `outputFileName`, and then the HTTP request will be used to resolve the name of the file.

Overall, it seems like that logic that decide of the fileName when not set in configuration should be moved inside this method.
The signature could then easily be `void download(URI uri, File targetFolder, String fileName, HttpClientContext clientContext)`

adamretter replied:
So it actually gets a bit more complex due to the caching of files. There was an issue in my code whereby you would have to set `<skipCache>true</skipCache>` to get things work run-after-run. I will update the PR with a further commit and explanation.

adamretter commented:
@mickaelistria Okay I added a second commit, which fixes an issue with caching and redirects. My use-case was this:

```xml
<plugin>
    <groupId>com.googlecode.maven-download-plugin</groupId>
    <artifactId>download-maven-plugin</artifactId>
    <configuration>
        <readTimeOut>120000</readTimeOut>
        <outputDirectory>${assemble.dir}/autodeploy</outputDirectory>
    </configuration>
    <executions>
        <execution>
            <id>autodeploy-dashboard-xar</id>
            <phase>package</phase>
            <goals>
                <goal>wget</goal>
            </goals>
            <configuration>
                <url>http://demo.exist-db.org/exist/apps/public-repo/find?processor=${project.version}&amp;abbrev=dashboard</url>
            </configuration>
        </execution>
    </executions>
</plugin>
```

The URL `http://demo.exist-db.org/exist/apps/public-repo/find?processor=5.0.0-SNAPSHOT&abbrev=dashboard` gets a response like this from the server:

```
< HTTP/1.1 302 Found
< Date: Fri, 12 Apr 2019 03:45:59 GMT
< Server: Jetty(9.4.9.v20180320)
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< X-XQuery-Cached: true
< Location: http://demo.exist-db.org/exist/apps/public-repo/public/dashboard-2.0.0-prod.xar
< Content-Length: 0
< Set-Cookie: JSESSIONID=node01e7v3ulg99okzzz5u3gc5o5h540441.node0;Path=/exist
```

## Before my second commit in this PR:
1. On the first call to the plugin, a request is made to the server, the response HTTP redirect would be automatically followed, the file downloaded from the server, and the file `dashboard-2.0.0-prod.xar` would be added to the cache `~/.m2/repository/.cache/download-maven-plugin/`.
2. On the second call to the plugin, the cache would be consulted for the file `find?processor=5.0.0-SNAPSHOT&abbrev=dashboard` (as that is the name in the original URI), the file would not be found. The HTTP redirect would be automatically followed again, the file would then be re-downloaded, and added to the cache a second time.

## After my second commit in this PR:
Redirects are not automatically followed. Upon a HTTP redirect, the cache is first consulted, if the file is in the cache the redirect is not followed (because we already have the file), otherwise the redirect is followed.

Hope my explanation makes sense? Happy to discuss...

mickaelistria commented:
Indeed, the overall implementation of cache sucks, but the file name isn't the key in cache, it's the URI. So whatever the file is named shouldn't affect cache resolution.
What's missing is the ability to have a cache entry be used for several URLs, so all the chain of redirect would be in cache, referencing the same file/bytes.

The approach of checking cache first before following a URL is good.

But as I'm thinking about usage of this issue, I'm wondering whether their is a good way to do this in general. Indeed, some redirects will starts with a "bad" URL which doesn't end with file name and redirect to a URL that end with file name, but some other redirects may do it the other way round (you ask for an URL that looks suffixed by a good file name and end up on some URL with generated UUIDs as name or something like that).
So I'm not sure we can really establish a strategy that covers redirect better in all cases, other strategies may be better for some cases, but not better in the general case.
I think redirects are a typical case where the user of the mojo should force themselves to use an `outputFileName`, since otherwise there is not much predictability and what the file name would be. I think it's probably not worth making the mojo more complex when there is already a recommended property that allows to workaround it very easily.
Maybe it'd be better to just "warn" when there is a redirect and one of the redirect URLs doesn't have a suffix that matches the file name, and recommend user that they specific the outputFileName in their pom instead to get more predictability.

adamretter commented:
> Indeed, the overall implementation of cache sucks, but the file name isn't the key in cache, it's the URI. 

Are you sure that is the case? Whenever I look in the `~/.m2/repository/.cache/download-maven-plugin/`, files are stored by filename and not URI.

> I think redirects are a typical case where the user of the mojo should force themselves to use an outputFileName, since otherwise there is not much predictability and what the file name would be. 

Unfortunately that isn't always possible, for example, my use-case, when you want to retrieve the latest version of a file, e.g. our URI - `http://demo.exist-db.org/exist/apps/public-repo/find?processor=5.0.0-SNAPSHOT&abbrev=dashboard`. Might return `dashboard-2.0.0.xar` today, but `dashboard-3.0.0.xar` tomorrow. We need to store the files by their version name, and we only know that from asking and downloading the latest file.

Does that make sense?

mickaelistria commented:
> Are you sure that is the case?

Yes, see the DownloadCache class.

> Might return dashboard-2.0.0.xar today, but dashboard-3.0.0.xar tomorrow

This is kind of dangerous to use anyway, because the cache will resolve to local dashboard-2.0.0.xar even if the URL now serve a different file.
What's the next step of your user-story: do you need the filename after fetch? Could you force a dashboard-latest.xar for instnace?

adamretter commented:
@michaelsembwever Yes exactly we need the correct filename after fetch.
We can actually "trust" the upstream server in this case ;-)

michaelsembwever commented:
@adamretter think you meant @mickaelistria :-)

adamretter commented:
@michaelsembwever Doh! ;-) Thanks.

adamretter commented:
>Might return dashboard-2.0.0.xar today, but dashboard-3.0.0.xar tomorrow

>This is kind of dangerous to use anyway, because the cache will resolve to local dashboard-2.0.0.xar even if the URL now serve a different file. What's the next step of your user-story: do you need the filename after fetch? Could you force a dashboard-latest.xar for instnace?

The entry in the cache will be for `dashboard-2.0.0.xar`, and the URI entry in the cache index is the URI of the redirected file (not the original URI) so it will never resolve for anything other than that file.

Even though this should not/would not happen... If someone starts delivering a different file content with the name `dashboard-2.0.0.xar` then it will have a different signature and won't be served from the cache.

So I think all is safe ;-)

mickaelistria commented:
So you change the signatures in the pom.xml? If so,why not changing the file name as well?

adamretter commented:
No the signatures are not in the pom

mickaelistria commented:
> the URI entry in the cache index is the URI of the redirected file (not the original URI)

I think this is not desired. If there is an entry in the cache for a URI, this entry should be used, even if the URI is redirected. The cache is also a way to work offline.

> If someone starts delivering a different file content with the name dashboard-2.0.0.xar then it will have a different signature and won't be served from the cache.
> [...]
> No the signatures are not in the pom

Signatures are not checked unless they are in the pom. If no signatures are specified, then cache is used anyway and the mojo will serve outdated content.

Basically, with redirects and without signatures, you have no guarantee at all that you get up-to-date content. It's an invalid workflow.
So to work well, either you skip the cache or you set signatures in pom files, and if you set signatures in pom file, you may as easily set a file name.
I'm starting to consider that the workflow you've set up is actually too dangerous to be implemented in the mojo. Cases with redirects and no file name specified and no signatures in pom files would probably better throw a warning or failure message to drive adopters to better usage, rather than building complexity in the mojo for usages that are essentially not reliable.
